### PR TITLE
[risk=no] DB=108

### PR DESCRIPTION
**Pull request by SrushtiGangireddy**
1. Added person_gender_identity mapping to calculate gender identity counts.
2. Added toggle function on gender chart to show biological sex / gender identity.
3. Help text for gender, age charts.

### CODE REVIEW

File: api/db-cdr/generate-cdr/bq-schemas/person_gender_identity.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[
```
Comment (calbach) [inline api/db-cdr/generate-cdr/bq-schemas/person_gender_identity.json]: This is an extra intermediate table we're writing to BigQuery, right?

File: api/db-cdr/generate-cdr/csv/achilles_analysis.csv [hunk 2]
```diff
@@ -6,9 +6,10 @@ analysis_id,"analysis_name","stratum_1_name","stratum_2_name","stratum_3_name","
 4,"Number of persons by race","race_concept_id","","","","","pie","counts"
 5,"Number of persons by ethnicity","ethnicity_concept_id","","","","","column","counts"
 6,"Number of persons by age decile","age_decile","","","","","column","counts"
+7,"Number of persons by gender_identity","gender_identity_concept_id","","","","","column","counts"
```
Comment (calbach) [inline api/db-cdr/generate-cdr/csv/achilles_analysis.csv]: Sorry, basic question, but what's the significance of the first column here? Is this just a database ID that you're choosing yourself (incrementally)?

File: api/db-cdr/generate-cdr/run-achilles-queries.sh [hunk 3]
```diff
@@ -184,6 +193,26 @@ COUNT(distinct p1.PERSON_ID) as count_value,COUNT(distinct p1.PERSON_ID) as sour
 on p1.person_id = co1.person_id where co1.condition_concept_id != co1.condition_source_concept_id
 group by co1.condition_source_concept_id, p1.gender_concept_id"
 
+# Condition gender identity
```
Comment (calbach) [inline api/db-cdr/generate-cdr/run-achilles-queries.sh]: Did you write the below SQL yourself or did it come from somewhere? Could you add a comment here summarizing what this is doing?

File: api/db-cdr/generate-cdr/run-achilles-queries.sh [hunk 4]
```diff
@@ -382,6 +411,30 @@ COUNT(distinct p1.PERSON_ID) as source_count_value from \`${BQ_PROJECT}.${BQ_DAT
 where co1.procedure_concept_id != co1.procedure_source_concept_id
 group by co1.procedure_source_concept_id,p1.gender_concept_id"
 
+#  3107 Gender identity
```
Comment (calbach) [inline api/db-cdr/generate-cdr/run-achilles-queries.sh]: How does this differ from the above query? COuld you add a comment to this one as well?

File: public-api/src/main/java/org/pmiops/workbench/cdr/model/QuestionConcept.java [hunk 5]
```diff
@@ -50,6 +54,16 @@ public static void setGenderStratumNameMap() {
         genderStratumNameMap.put("8570", "Ambiguous");
     }
 
+    public static void setGenderIdentityStratumNameMap() {
+        genderIdentityStratumNameMap.put("1585840", "Woman");
```
Comment (calbach) [inline public-api/src/main/java/org/pmiops/workbench/cdr/model/QuestionConcept.java]: Are these concept IDs? Where is this mapping originally defined/documneted?

File: public-ui/src/app/data-browser/chart/chart.component.ts [hunk 6]
```diff
@@ -81,11 +81,22 @@ export class ChartComponent implements OnChanges {
         pie: {
           borderColor: null,
           slicedOffset: 4,
+          size: this.analysis ?
+              (this.analysis.analysisId === this.dbc.SURVEY_GENDER_IDENTITY_ANALYSIS_ID ||
```
Comment (calbach) [inline public-ui/src/app/data-browser/chart/chart.component.ts]: For these 3 cases I would probably pull this boolean (or two booleans?) out into a descriptively named local varaible, I'm not following what you are checking for here:

```
const isSurveyGenderAnalysis = ...;
...
size: isSurveryGenderAnalysis ? '60%' : '100%';
```

File: public-ui/src/app/data-browser/chart/chart.component.ts [hunk 7]
```diff
@@ -296,6 +313,7 @@ export class ChartComponent implements OnChanges {
       // Series name for answers is the answer selected which is in stratum4
       seriesName = this.selectedResult.stratum4;
     }
+    console.log(results);
```
Comment (calbach) [inline public-ui/src/app/data-browser/chart/chart.component.ts]: Assuming this was left in from debugging, please remove

File: public-ui/src/app/data-browser/concept-charts/concept-charts.component.html [hunk 8]
```diff
@@ -6,12 +6,117 @@
...
+    </clr-tooltip-content>
+  </clr-tooltip>
 </div>
 
 <div *ngIf="!loading()" class="row concept-charts chart">
   <!-- show measurements binned by gender if we want -->
   <!-- Loop through the genders we have measurement results for  for showing the graph
    This looping variable is set if showMeasurements is passed in to the component-->
+  <ng-container *ngIf="(showGender || showGenderIdentity) && showAge && (showSources && sourceConcepts.length > 0) && !showMeasurementGenderBins">
+  <div class="col-md-4 col-xs-12">
```
Comment (calbach) [inline public-ui/src/app/data-browser/concept-charts/concept-charts.component.html]: nit: indent?

File: public-ui/src/app/data-browser/concept-charts/concept-charts.component.html [hunk 9]
```diff
@@ -6,12 +6,117 @@
...
+      <div class="tooltip-label">Toggle here to view unit specific histograms</div>
+    </clr-tooltip-content>
+  </clr-tooltip>
 </div>
 
 <div *ngIf="!loading()" class="row concept-charts chart">
   <!-- show measurements binned by gender if we want -->
   <!-- Loop through the genders we have measurement results for  for showing the graph
    This looping variable is set if showMeasurements is passed in to the component-->
+  <ng-container *ngIf="(showGender || showGenderIdentity) && showAge && (showSources && sourceConcepts.length > 0) && !showMeasurementGenderBins">
```
Comment (calbach) [inline public-ui/src/app/data-browser/concept-charts/concept-charts.component.html]: This is a pretty complex predicate. I would recommend putting this int the component.ts as a method and calling that here, e.g. `showX` (not sure what X is here.

File: public-api/src/main/java/org/pmiops/workbench/cdr/model/QuestionConcept.java (hunk 5 above)
Comment (SrushtiGangireddy) [reply public-api/src/main/java/org/pmiops/workbench/cdr/model/QuestionConcept.java]: i made this mapping myself, looking at the distinct answers to the question and their names in concept table.

File: api/db-cdr/generate-cdr/run-achilles-queries.sh (hunk 4 above)
Comment (SrushtiGangireddy) [reply api/db-cdr/generate-cdr/run-achilles-queries.sh]: this calculates the gender identity counts for procedure concepts.

File: api/db-cdr/generate-cdr/run-achilles-queries.sh (hunk 3 above)
Comment (SrushtiGangireddy) [reply api/db-cdr/generate-cdr/run-achilles-queries.sh]: i wrote this myself, this is similar to the one above (3101) except that this calculates the count of gender identity from the table created above unlike the gender counts calculated in 3101 from person bq table.

File: api/db-cdr/generate-cdr/csv/achilles_analysis.csv (hunk 2 above)
Comment (SrushtiGangireddy) [reply api/db-cdr/generate-cdr/csv/achilles_analysis.csv]: yeah, the auto id we are just adding when new row is added

File: api/db-cdr/generate-cdr/bq-schemas/person_gender_identity.json (hunk 1 above)
Comment (SrushtiGangireddy) [reply api/db-cdr/generate-cdr/bq-schemas/person_gender_identity.json]: yeah, so basically we want to also show the gender identity graph along with the biological sex pie.

File: public-ui/src/app/data-browser/chart/chart.component.ts (hunk 6 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/data-browser/chart/chart.component.ts]: The gender charts need a toggle button on the top which is making the gender chart overflow in to age so was trying to reduce the height in that case.

File: public-ui/src/app/data-browser/chart/chart.component.ts (hunk 7 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/data-browser/chart/chart.component.ts]: Done.

File: public-ui/src/app/data-browser/concept-charts/concept-charts.component.html (hunk 8 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/data-browser/concept-charts/concept-charts.component.html]: Done.

File: public-ui/src/app/data-browser/concept-charts/concept-charts.component.html (hunk 9 above)
Comment (SrushtiGangireddy) [reply public-ui/src/app/data-browser/concept-charts/concept-charts.component.html]: Moved it into a method in .ts

File: api/db-cdr/generate-cdr/run-achilles-queries.sh [hunk 10]
```diff
@@ -184,6 +193,27 @@ COUNT(distinct p1.PERSON_ID) as count_value,COUNT(distinct p1.PERSON_ID) as sour
...
+# Condition gender identity
+# Calculating the gender identity counts for condition concepts based on the answer of each person to ppi question 1585838
+bq --quiet --project=$BQ_PROJECT query --nouse_legacy_sql \
+"insert into \`${WORKBENCH_PROJECT}.${WORKBENCH_DATASET}.achilles_results\`
+(id, analysis_id, stratum_1, stratum_2, stratum_3, count_value, source_count_value)
+select 0, 3107 as analysis_id,
+	CAST(co1.condition_concept_id AS STRING) as stratum_1,
+	CAST(p1.gender_identity_concept_id AS STRING) as stratum_2,'Condition' as stratum_3,
+	COUNT(distinct p1.PERSON_ID) as count_value,
+	(select COUNT(distinct p2.PERSON_ID) from \`${WORKBENCH_PROJECT}.${WORKBENCH_DATASET}.person_gender_identity\` p2 inner join \`${BQ_PROJECT}.${BQ_DATASET}.condition_occurrence\` co2 on p2.person_id=co2.person_id where co2.condition_source_concept_id=co1.condition_concept_id and p2.gender_identity_concept_id = p1.gender_identity_concept_id) as source_count_value
```
Comment (calbach) [inline api/db-cdr/generate-cdr/run-achilles-queries.sh]: nit: can this line be broken up and indented (here and below)? This is difficult to read

File: api/db-cdr/generate-cdr/run-achilles-queries.sh [hunk 11]
```diff
@@ -98,6 +99,14 @@ from \`${BQ_PROJECT}.${BQ_DATASET}.person\`
 where floor((2018 - year_of_birth)/10) >=3
 group by stratum_2"
 
+# 7 Gender identity count
```
Comment (calbach) [inline api/db-cdr/generate-cdr/run-achilles-queries.sh]: The content here definitely feels like it's grown well beyond the limits of a shell script. There's a pretty high degree of duplication in this file and it is quite difficult to review.

Not for this PR, but possibly a future refactoring ticket.

File: api/db-cdr/generate-cdr/run-achilles-queries.sh (hunk 10 above)
Comment (SrushtiGangireddy) [reply api/db-cdr/generate-cdr/run-achilles-queries.sh]: Alright, pushed in the change. Thanks.

File: api/db-cdr/generate-cdr/run-achilles-queries.sh (hunk 11 above)
Comment (SrushtiGangireddy) [reply api/db-cdr/generate-cdr/run-achilles-queries.sh]: I was thinking about that too. Has been adding up things lately. Will take care of this in next pr.

**Pull request merged**